src/integrations.py

```diff
@@ -98,3 +120,132 @@ def auxiliary_data(self) -> AuxiliaryData:
             username=database_config.username,
             password=database_config.password,
         )
+
+
+@dataclass
+class CertificateData:
+    ca_cert: Optional[str] = None
+    ca_chain: Optional[list[str]] = None
+    cert: Optional[str] = None
+
+
+class CertificatesIntegration:
+    def __init__(self, charm: CharmBase) -> None:
+        self._charm = charm
+        self._container = charm._container
+
+        hostname = charm.config.get("hostname")
+        self.cert_handler = CertHandler(
+            charm,
+            key="glauth-server-cert",
+            peer_relation_name="glauth-peers",
+            cert_subject=hostname,
+            extra_sans_dns=[hostname],
+        )
+
+    @property
+    def _ca_cert(self) -> Optional[str]:
+        return self.cert_handler.ca
+
+    @property
+    def _server_key(self) -> Optional[str]:
+        return self.cert_handler.key
+
+    @property
+    def _server_cert(self) -> Optional[str]:
+        return self.cert_handler.cert
+
+    @property
+    def _ca_chain(self) -> list[str]:
+        return self.cert_handler.chain
+
+    @property
+    def cert_data(self) -> CertificateData:
+        return CertificateData(
+            ca_cert=self._ca_cert,
+            ca_chain=self._ca_chain,
+            cert=self._server_cert,
+        )
+
+    def update_certificates(self) -> None:
+        if not self.cert_handler.enabled:
+            logger.debug("The certificates integration is not ready.")
+            self._remove_certificates()
+            return
+
+        if not self.certs_ready():
+            logger.debug("The certificates data is not ready.")
+            self._remove_certificates()
+            return
+
+        self._prepare_certificates()
+        self._add_certificates()
+
+    def certs_ready(self) -> bool:
+        return all((self._ca_cert, self._ca_chain, self._server_key, self._server_cert))
+
+    def _prepare_certificates(self) -> None:
+        SERVER_CA_CERT.write_text(self._ca_cert)  # type: ignore[arg-type]
+        SERVER_KEY.write_text(self._server_key)  # type: ignore[arg-type]
```

Comment:
Correct me if I'm wrong, but the charm container does not need to know the server private key.

Reply:
It needs the private key to be present.